Content.Shared/Chameleon/SharedChameleonSystem.cs

```diff
@@ -19,46 +15,95 @@ public override void Initialize()
         SubscribeLocalEvent<SharedChameleonComponent, ComponentGetState>(OnChameleonGetState);
         SubscribeLocalEvent<SharedChameleonComponent, ComponentHandleState>(OnChameleonHandlesState);
         SubscribeLocalEvent<SharedChameleonComponent, MoveEvent>(OnMove);
+        SubscribeLocalEvent<SharedChameleonComponent, EntityPausedEvent>(OnPaused);
+        SubscribeLocalEvent<SharedChameleonComponent, ComponentInit>(OnInit);
     }
 
-    public override void Update(float frameTime)
+    private void OnPaused(EntityUid uid, SharedChameleonComponent component, EntityPausedEvent args)
     {
-        base.Update(frameTime);
-
-        foreach (var chameleon in EntityQuery<SharedChameleonComponent>())
+        if (args.Paused)
+        {
+            component.LastVisibility = Getvisibility(uid, component);
+            component.LastUpdated = null;
+        }
+        else
         {
-            //Only update if the current stealth level is above the stealth threshold
-            //This stops it from needlessly updating and calling Dirty once it's hit max stealth
-            if (chameleon.StealthLevel > StealthThreshold)
-            {
-                chameleon.StealthLevel = Math.Clamp(chameleon.StealthLevel - frameTime * chameleon.InvisibilityRate, -1f, 1f);
-
-                Dirty(chameleon);
-            }
+            component.LastUpdated = _timing.CurTime;
         }
+
+        Dirty(component);
+    }
+
+    protected virtual void OnInit(EntityUid uid, SharedChameleonComponent component, ComponentInit args)
+    {
+        if (component.LastUpdated != null || Paused(uid))
+            return;
+
+        component.LastUpdated = _timing.CurTime;
     }
 
     private void OnChameleonGetState(EntityUid uid, SharedChameleonComponent component, ref ComponentGetState args)
     {
-        args.State = new ChameleonComponentState(component.StealthLevel);
+        args.State = new ChameleonComponentState(component.LastVisibility, component.LastUpdated);
     }
 
     private void OnChameleonHandlesState(EntityUid uid, SharedChameleonComponent component, ref ComponentHandleState args)
     {
         if (args.Current is not ChameleonComponentState cast)
             return;
 
-        component.StealthLevel = cast.StealthLevel;
+        component.LastVisibility = cast.Visibility;
+        component.LastUpdated = cast.LastUpdated;
     }
 
     private void OnMove(EntityUid uid, SharedChameleonComponent component, ref MoveEvent args)
     {
+        if (args.FromStateHandling)
+            return;
+
```

Comment:
And this was an oversight of mine when it comes to state handling. In particular, if an entity's transform component state gets applied after the chameleon state, this would incorrectly modify the current stealth level.